### PR TITLE
Update project to Xcode 8

### PR DESCRIPTION
**Pull request by eliperkins**
Hey y'all. Wanted to start the ball rolling on the Xcode 8 train. 🚂🚋🚋🚋🚋

I updated the project to Xcode 8 and the latest Swift 3 syntax.

The big change here is appeasing Xcode's warning about single parameter functions needing to be enclosed in parens. That is:

``` swift
// Warning: Single argument function types require parentheses.
public func curry<A, B, C>(function: (A, B) -> C) -> A -> B -> C {
...
}

// Does not warn
public func curry<A, B, C>(function: (A, B) -> C) -> (A) -> (B) -> C {
...
}
```

Also touched the generator script to be Swift 3 compatible.

The generated `Curry.swift` is backwards compatible with Swift 2, while the generator script is not.

Lemme know if this should be pointed at a new `swift-3` or `xcode-8` branch or whatever! Figured I'd open the PR and start the discussion here.

### CODE REVIEW

File: bin/generate [hunk 1]
```diff
@@ -8,48 +8,56 @@ let generics = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M",
 
 extension Array {
   subscript(safe index: Int) -> Element? {
-    return indices ~= index ? self[index] : .None
+    return indices ~= index ? self[index] : .none
   }
 }
 
-func genericTypeForPosition(x: Int) -> String {
+func genericType(for position: Int) -> String {
```
Comment (gfontenot) [inline bin/generate:15]: nice

File: bin/generate [hunk 2]
```diff
@@ -8,48 +8,56 @@ let generics = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M",
...
 }
 
-func curryDefinitionGenerator(arguments arguments: Int) -> String {
-  let genericParameters = (0..<arguments).map(genericTypeForPosition) // ["A", "B", "C"]
-  let genericTypeDefinition = "<\(commaConcat(genericParameters))>" // "<A, B, C>"
+func singleParameterFunctions(_ xs: [String]) -> String {
+  guard let first = xs.first else { fatalError("Attempted to nest functions with no arguments") }
+  guard xs.last != first else { return first }
+  let remainder = Array(xs.dropFirst())
+  return "(\(first)) -> \(singleParameterFunctions(remainder))"
```
Comment (gfontenot) [inline bin/generate:31]: oooohhhh recursion

File: bin/generate [hunk 3]
```diff
@@ -8,48 +8,56 @@ let generics = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M",
...
 
-  let functionArguments = "(\(commaConcat(inputParameters)))" // "(A, B)"
-  let returnFunctionArguments = inputParameters.joinWithSeparator(" -> ") // " A -> B"
-  let innerFunctionArguments = commaConcat(lowerFunctionArguments) // "`a`, `b`"
+  let functionArguments = "(\(commaConcat(inputParameters)))" // "(A, B, C)"
+  let returnFunction = singleParameterFunctions(genericParameters) // " (A) -> (B) -> (C) -> D"
+  let innerFunctionArguments = commaConcat(lowerFunctionArguments) // "`a`, `b`, `c`"
 
-  let functionDefinition = "function(\(innerFunctionArguments))" // return function(`a`, `b`)
+  let functionDefinition = "function(\(innerFunctionArguments))" // return function(`a`, `b`, `c`)
```
Comment (gfontenot) [inline bin/generate]: Can we update this code comment to remove `"return"`? Doesn't look like it's actually added here.

**Comment by gfontenot** [comment]
One actionable comment, but this seems reasonable to me. Given that it's source compatible, I don't see any reason not to move forward with it.

**Comment by eliperkins** [comment]
🎊 Thanks @gfontenot!

**Comment by gfontenot** [comment]
✨ ✨ ✨ ✨ 

Released as v2.3.0

**Pull request merged**